compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java

```diff
@@ -813,24 +813,32 @@ public void visit(BLangMatchClause matchClause) {
     public void visit(BLangForeach foreach) {
         analyzeNode(foreach.collection, env);
         analyzeNode(foreach.body, env);
+
+        BLangOnFailClause onFailClause = foreach.onFailClause;
+        if (onFailClause != null) {
+            analyzeNode(onFailClause, env);
+        }
     }
 
     @Override
     public void visit(BLangFromClause fromClause) {
-        analyzeNode((BLangNode) fromClause.getVariableDefinitionNode(), env);
-        analyzeNode(fromClause.collection, env);
+        SymbolEnv fromEnv = fromClause.env;
+        analyzeNode((BLangNode) fromClause.getVariableDefinitionNode(), fromEnv);
+        analyzeNode(fromClause.collection, fromEnv);
     }
 
     @Override
     public void visit(BLangJoinClause joinClause) {
-        analyzeNode((BLangNode) joinClause.getVariableDefinitionNode(), env);
-        analyzeNode(joinClause.collection, env);
+        SymbolEnv joinEnv = joinClause.env;
+        analyzeNode((BLangNode) joinClause.getVariableDefinitionNode(), joinEnv);
+        analyzeNode(joinClause.collection, joinEnv);
```

Comment:
```suggestion
        analyzeNode(joinClause.collection, joinEnv);
        analyzeNode((BLangNode) joinClause.onClause, joinEnv);
```

Reply:
Thanks, had missed this. Fixed and added tests in 8dc7bef.